src/util/log/fd_log.c

```diff
@@ -1455,51 +1455,78 @@ fd_log_private_stack_discover( ulong   stack_sz,
   FD_VOLATILE( stack_mem[0] ) = (uchar)1; /* Paranoia to make sure compiler puts this in stack */
   ulong stack_addr = (ulong)stack_mem;
 
-  FILE * file = fopen( "/proc/self/maps", "r" );
-  if( FD_UNLIKELY( !file ) )
-    FD_LOG_WARNING(( "fopen( \"/proc/self/maps\", \"r\" ) failed (%i-%s)", errno, fd_io_strerror( errno ) ));
-  else {
-
-    while( FD_LIKELY( !feof( file ) ) ) {
-
-      /* Get the next memory region */
-
-      char buf[ 1024 ];
-      if( FD_UNLIKELY( !fgets( buf, 1024UL, file ) ) ) break;
-      ulong m0;
-      ulong m1;
-      if( FD_UNLIKELY( sscanf( buf, "%lx-%lx", &m0, &m1 )!=2 ) ) continue;
-
-      /* Test if the stack allocation is in the discovered region */
-
-      if( FD_UNLIKELY( (m0<=stack_addr) & (stack_addr<m1) ) ) {
-        ulong msz = m1 - m0;
-        if( msz==stack_sz ) { /* Memory region matches expectations */
-          stack0 = m0;
-          stack1 = m1;
-        } else if( ((fd_log_group_id()==fd_log_tid()) & (msz<stack_sz)) ) {
-          /* This is the main thread, which, on recent Linux, seems to
-             just reserve address space for main's stack at program
-             start up to the application stack size limits then uses
-             page faults to dynamically back the stack with DRAM as the
-             stack grows (which is awful for performance, jitter and
-             reliability ... sigh).  This assumes stack grows down such
-             that m1 is the fixed value in this process. */
-          stack0 = m1 - stack_sz;
-          stack1 = m1;
-        } else {
-          FD_LOG_WARNING(( "unexpected caller stack memory region size (got %lu bytes, expected %lu bytes)", msz, stack_sz ));
-          /* don't trust the discovered region */
-        }
-        break;
-      }
+  int filefd;
+  if( FD_UNLIKELY( ( filefd = open( "/proc/self/maps", O_RDONLY ) ) < 0 ) ) {
+    FD_LOG_WARNING(( "open( \"/proc/self/maps\" ) failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+    *_stack0 = 0UL;
+    *_stack1 = 0UL;
+    return;
+  }
 
+  char filebuf[1<<16];
```

Comment:
I don't love the change here to read the whole file up-front, since there's no guarantee the maps listing fits in this buffer. Can we keep the old line-based parsing?

Reply:
that's complicated because we might read partial lines

Reply:
The implementation already did streaming? You can just leave it basically as it was, except swap fread for read. Line length is guaranteed to be a certain size, since the file format is fixed; while the number of lines is unlimited.

Also this code is in util so probably shouldn't abort or do other shortcuts like that. You also might want to keep the FD_LOG_WARNING if close() fails.

Reply:
@ripatel-fd @mmcgee-jump the problem is that I need to read a line at a time, and I don't know how long the lines are. This requires a buffering/push back solution. raw read and fd_io_buffered doesn't support that

Reply:
What was wrong with the existing impl? Also you do know how long the lines are... the file format is fixed, the first 5 fields are fixed width, and the last (pathname) is bounded by PATH_MAX.

Reply:
fopen uses a hidden malloc which was crashing. the problem is that if you are reading fixed size chunks, you'll get partial lines, which then have to be shifted over for the next read. that is too much code for what is a trivial application

Reply:
that's a mystery

Reply:
I think it's worth adding the code to shift the buffer here -- this code is in util so has a higher quality bar than other parts of the codebase, is intended to be reused by other arbitrary non-Firedancer callers.  If you don't do the line-by-line parser you arbitrarily cap the number of mapped regions a process can have where this code works to ... approx ~750. (After which it will fail silently except for an inscrutable warning message and let the caller proceed to crash later.)

Reply:
done